File: packages/agentmark-core/src/schemas.ts
import { z } from "zod";

export const ChatMessageSchema = z.object({
  role: z.enum(["system", "user", "assistant"]),
  content: z.string(),
});

export type ChatMessage = z.infer<typeof ChatMessageSchema>;

export const textPartSchema = z.object({
  type: z.literal("text"),
  text: z.string(),
});

export const imagePartSchema = z.object({
  type: z.literal("image"),
  image: z.string().url(),
  mimeType: z.string().optional(),
});

export const filePartSchema = z.object({
  type: z.literal("file"),
  data: z.string().url(),
  mimeType: z.string(),
});

export const userMessagesSchema = z.object({
  role: z.literal("user"),
  content: z.union([
    z.string(),
    z.array(z.union([textPartSchema, imagePartSchema, filePartSchema])),
  ]),
});

export const assistantMessagesSchema = z.object({
  role: z.literal("assistant"),
  content: z.string(),
});

export const systemMessagesSchema = z.object({
  role: z.literal("system"),
  content: z.string(),
});

export const RichChatMessageSchema = z.union([
  userMessagesSchema,
  assistantMessagesSchema,
  systemMessagesSchema,
]);

export type RichChatMessage = z.infer<typeof RichChatMessageSchema>;

export const TextSettingsConfig = z.object({
  model_name: z.string(),
  max_tokens: z.number().optional(),
  temperature: z.number().optional(),
  max_calls: z.number().optional(),
  top_p: z.number().optional(),
  top_k: z.number().optional(),
  presence_penalty: z.number().optional(),
  frequency_penalty: z.number().optional(),
  stop_sequences: z.array(z.string()).optional(),
  seed: z.number().optional(),
  max_retries: z.number().optional(),
  tool_choice: z
    .union([
      z.enum(["auto", "none", "required"]),
      z.object({
        type: z.literal("tool"),
        tool_name: z.string(),
      }),
    ])
    .optional(),
  tools: z
    .record(
      z.object({
        description: z.string(),
        parameters: z.record(z.any()),
      })
    )
    .optional(),
});

export type TextSettings = z.infer<typeof TextSettingsConfig>;

export const TestSettingsSchema = z.object({
  props: z.record(z.any()).nullable().optional(),
  dataset: z.string().optional(),
});

export type TestSettings = z.infer<typeof TestSettingsSchema>;

export const ObjectSettingsConfig = z.object({
  model_name: z.string(),
  max_tokens: z.number().optional(),
  temperature: z.number().optional(),
  max_calls: z.number().optional(),
  top_p: z.number().optional(),
  top_k: z.number().optional(),
  presence_penalty: z.number().optional(),
  frequency_penalty: z.number().optional(),
  stop_sequences: z.array(z.string()).optional(),
  seed: z.number().optional(),
  max_retries: z.number().optional(),
  schema: z.record(z.any()),
  schema_name: z.string().optional(),
  schema_description: z.string().optional(),
});

export type ObjectSettings = z.infer<typeof ObjectSettingsConfig>;

export const ImageSettingsConfig = z.object({
  model_name: z.string(),
  num_images: z.number().optional(),
  size: z
    .string()
    .regex(/^\d+x\d+$/)
    .optional(),
  aspect_ratio: z
    .string()
    .regex(/^\d+:\d+$/)
    .optional(),
  seed: z.number().optional(),
});

export type ImageSettings = z.infer<typeof ImageSettingsConfig>;

export const SpeechSettingsConfig = z.object({
  model_name: z.string(),
  voice: z.string().optional(),
  output_format: z.string().optional(),
  speed: z.number().optional(),
});

export type SpeechSettings = z.infer<typeof SpeechSettingsConfig>;

export const TextConfigSchema = z.object({
  name: z.string(),
  messages: z.array(RichChatMessageSchema),
  text_config: TextSettingsConfig,
  test_settings: TestSettingsSchema.optional(),
});

export type TextConfig = z.infer<typeof TextConfigSchema>;

export const ObjectConfigSchema = z.object({
  name: z.string(),
  messages: z.array(RichChatMessageSchema),
  object_config: ObjectSettingsConfig,
  test_settings: TestSettingsSchema.optional(),
});

export type ObjectConfig = z.infer<typeof ObjectConfigSchema>;

export const ImageConfigSchema = z.object({
  name: z.string(),
  image_config: ImageSettingsConfig,
  test_settings: TestSettingsSchema.optional(),
});

export type ImageConfig = z.infer<typeof ImageConfigSchema>;

export const SpeechConfigSchema = z.object({
  name: z.string(),
  speech_config: SpeechSettingsConfig,
  test_settings: TestSettingsSchema.optional(),
});

export type SpeechConfig = z.infer<typeof SpeechConfigSchema>;

File: packages/prompt-core-python/src/agentmark/prompt_core/schemas.py
"""Pydantic schemas for prompt configurations."""

from typing import Annotated, Any, Literal

from pydantic import BaseModel, ConfigDict, Field


class TextPartSchema(BaseModel):
    """Text content part schema."""

    model_config = ConfigDict(extra="forbid")

    type: Literal["text"]
    text: str


class ImagePartSchema(BaseModel):
    """Image content part schema."""

    model_config = ConfigDict(extra="forbid")

    type: Literal["image"]
    image: str
    mimeType: str | None = None


class FilePartSchema(BaseModel):
    """File content part schema."""

    model_config = ConfigDict(extra="forbid")

    type: Literal["file"]
    data: str
    mimeType: str


ContentPartSchema = TextPartSchema | ImagePartSchema | FilePartSchema


class UserMessageSchema(BaseModel):
    """User message schema."""

    model_config = ConfigDict(extra="forbid")

    role: Literal["user"]
    content: str | list[ContentPartSchema]


class AssistantMessageSchema(BaseModel):
    """Assistant message schema."""

    model_config = ConfigDict(extra="forbid")

    role: Literal["assistant"]
    content: str


class SystemMessageSchema(BaseModel):
    """System message schema."""

    model_config = ConfigDict(extra="forbid")

    role: Literal["system"]
    content: str


RichChatMessageSchema = UserMessageSchema | AssistantMessageSchema | SystemMessageSchema


class ToolChoiceSchema(BaseModel):
    """Tool choice schema."""

    model_config = ConfigDict(extra="forbid")

    type: Literal["tool"]
    tool_name: str


class TextSettingsSchema(BaseModel):
    """Text generation settings schema."""

    model_config = ConfigDict(extra="forbid")

    model_name: str
    max_tokens: int | None = None
    temperature: float | None = None
    max_calls: int | None = None
    top_p: float | None = None
    top_k: int | None = None
    presence_penalty: float | None = None
    frequency_penalty: float | None = None
    stop_sequences: list[str] | None = None
    seed: int | None = None
    max_retries: int | None = None
    tool_choice: Literal["auto", "none", "required"] | ToolChoiceSchema | None = None
    tools: dict[str, str | dict[str, Any]] | None = None


class ObjectSettingsSchema(BaseModel):
    """Object/structured output settings schema."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    model_name: str
    max_tokens: int | None = None
    temperature: float | None = None
    max_calls: int | None = None
    top_p: float | None = None
    top_k: int | None = None
    presence_penalty: float | None = None
    frequency_penalty: float | None = None
    stop_sequences: list[str] | None = None
    seed: int | None = None
    max_retries: int | None = None
    schema_: Annotated[dict[str, Any], Field(alias="schema")]
    schema_name: str | None = None
    schema_description: str | None = None


class ImageSettingsSchema(BaseModel):
    """Image generation settings schema."""

    model_config = ConfigDict(extra="forbid")

    model_name: str
    prompt: str = ""
    num_images: int | None = None
    size: str | None = None  # e.g., "1024x1024"
    aspect_ratio: str | None = None  # e.g., "1:1"
    seed: int | None = None


class SpeechSettingsSchema(BaseModel):
    """Speech synthesis settings schema."""

    model_config = ConfigDict(extra="forbid")

    model_name: str
    text: str = ""
    voice: str | None = None
    output_format: str | None = None
    instructions: str | None = None
    speed: float | None = None


class TestSettingsSchema(BaseModel):
    """Test settings schema."""

    model_config = ConfigDict(extra="forbid")

    props: dict[str, Any] | None = None
    dataset: str | None = None
    evals: list[str] | None = None


class TextConfigSchema(BaseModel):
    """Text prompt configuration schema."""

    model_config = ConfigDict(extra="forbid")

    name: str
    messages: list[RichChatMessageSchema]
    text_config: TextSettingsSchema
    test_settings: TestSettingsSchema | None = None
    agentmark_meta: dict[str, Any] | None = None


class ObjectConfigSchema(BaseModel):
    """Object prompt configuration schema."""

    model_config = ConfigDict(extra="forbid")

    name: str
    messages: list[RichChatMessageSchema]
    object_config: ObjectSettingsSchema
    test_settings: TestSettingsSchema | None = None
    agentmark_meta: dict[str, Any] | None = None


class ImageConfigSchema(BaseModel):
    """Image prompt configuration schema."""

    model_config = ConfigDict(extra="forbid")

    name: str
    image_config: ImageSettingsSchema
    test_settings: TestSettingsSchema | None = None
    agentmark_meta: dict[str, Any] | None = None


class SpeechConfigSchema(BaseModel):
    """Speech prompt configuration schema."""

    model_config = ConfigDict(extra="forbid")

    name: str
    speech_config: SpeechSettingsSchema
    test_settings: TestSettingsSchema | None = None
    agentmark_meta: dict[str, Any] | None = None


AgentmarkConfigSchema = TextConfigSchema | ObjectConfigSchema | ImageConfigSchema | SpeechConfigSchema
